modules/pairIdSystem.js

```diff
@@ -0,0 +1,77 @@
+/**
+ * This module adds PAIR Id to the User ID module
+ * The {@link module:modules/userId} module is required
+ * @module modules/pairIdSystem
+ * @requires module:modules/userId
+ */
+
+import { submodule } from '../src/hook.js';
+import {getStorageManager} from '../src/storageManager.js'
+import { logError } from '../src/utils.js';
+
+const MODULE_NAME = 'pairId';
+const PAIR_ID_KEY = 'pairId';
+const LIVERAMP_PAIR_ID_KEY = '_lr_pairId';
+
+export const storage = getStorageManager()
+
+function pairIdFromLocalStorage(key) {
+  return storage.localStorageIsEnabled ? storage.getDataFromLocalStorage(key) : null;
+}
+
+function pairIdFromCookie(key) {
+  return storage.cookiesAreEnabled ? storage.getCookie(key) : null;
+}
+
+/** @type {Submodule} */
+export const pairIdSubmodule = {
+  /**
+  * used to link submodule with config
+  * @type {string}
+  */
+  name: MODULE_NAME,
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function
+   * @param { string | undefined } value
+   * @returns {{pairId:string} | undefined }
+   */
+  decode(value) {
+    return value && Array.isArray(value) ? {'pairId': value} : undefined
+  },
+  /**
+  * performs action to obtain id and return a value in the callback's response argument
+  * @function
+  * @returns {id: string | undefined }
+  */
+  getId(config) {
+    const pairIdsString = pairIdFromLocalStorage(PAIR_ID_KEY) || pairIdFromCookie(PAIR_ID_KEY)
+    let ids = []
+    if (pairIdsString && typeof pairIdsString == 'string') {
+      try {
+        ids = ids.concat(JSON.parse(atob(pairIdsString)))
+      } catch (error) {
+        logError(error)
+      }
+    }
+
+    const configParams = (config && config.params) || {};
+    if (configParams && configParams.liveramp) {
+      const liverampValue = pairIdFromLocalStorage(LIVERAMP_PAIR_ID_KEY) || pairIdFromCookie(LIVERAMP_PAIR_ID_KEY)
```

Comment:
Okay, on second pass and understanding this is not a liveramp contribution, i believe I understand this a lot better. Can you add the LIVERAMP_PAIR_ID_KEY as a config param  (yet falling back to the default string you have above) and loop over all the configured params?

My goal here is that a publisher would not need to upgrade the js library as support for new pair keys and partners are added, but could instead specify everything in config.

Reply:
I can have LIVERAMP_PAIR_ID_KEY as a config param, but probably won't serve the user much, as it's an agreed on value between us and liveramp, and they are fetching IDs and putting them into LIVERAMP_PAIR_ID_KEY with liveramp scripts they provide to the publishers.

As for supporting more partners through config, I understand this question, but this is a non-goal for us at the moment. Because: 1) we are focusing on enabling MVP customer testing, partnering with liveramp. 2) For the other partners, i'm not sure they will follow the same pattern, probably they would prefer the publishers to do a call to their API endpoint. So other pairId provider will still need to come and provide their logic themselves, like a separate ID provider. We will not dictate how they provide PairId to publishers.

Reply:
Sounds reasonable